crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+        count: u32,
+        index: u32,
+        parents: &[ComponentState],
+        offset: usize,
+    ) -> Result<()> {
+        let count = count as usize;
+        let ty = if count == 0 {
+            self.get(index, offset)?.clone()
+        } else {
+            if count > parents.len() {
+                return Err(BinaryReaderError::new(
+                    format!("invalid outer alias count of {}", count),
+                    offset,
+                ));
+            }
+
+            parents[parents.len() - count]
+                .types
+                .get(index, offset)?
+                .clone()
+        };
+
+        self.0.push(ty);
+
+        Ok(())
+    }
+
+    fn module_type(
+        defs: &[crate::ModuleType],
+        features: &WasmFeatures,
+        offset: usize,
+    ) -> Result<ModuleType> {
+        let mut types = ModuleTypeSpace::default();
+        let mut imports = HashMap::new();
+        let mut exports = HashMap::new();
+
+        for def in defs {
+            match def {
+                crate::ModuleType::Type(ty) => {
+                    types.type_def(ty.clone(), features, offset)?;
+                }
+                crate::ModuleType::Export { name, ty } => {
+                    let ty = types.check_type_ref(ty, features, offset)?;
+
+                    if exports.insert(name.to_string(), ty).is_some() {
+                        return Err(BinaryReaderError::new(
+                            format!("duplicate export name `{}` already defined", name),
+                            offset,
+                        ));
+                    }
+                }
+                crate::ModuleType::Import(i) => {
+                    let ty = types.check_type_ref(&i.ty, features, offset)?;
+
+                    if imports
+                        .insert((i.module.to_string(), i.name.to_string()), ty)
+                        .is_some()
+                    {
+                        return Err(BinaryReaderError::new(
+                            format!(
+                                "duplicate import name `{}::{}` already defined",
+                                i.module, i.name
+                            ),
+                            offset,
+                        ));
+                    }
+                }
+            };
+        }
+
+        Ok(ModuleType {
+            imports: Rc::new(imports),
+            exports: Rc::new(exports),
+        })
+    }
+
+    fn component_type(
+        defs: &[crate::ComponentType],
+        features: &WasmFeatures,
+        parents: &[ComponentState],
+        offset: usize,
+    ) -> Result<ComponentType> {
+        let mut types = ComponentTypeSpace::default();
+        let mut imports = HashMap::new();
+        let mut exports = HashMap::new();
+
+        for def in defs {
+            match def {
+                crate::ComponentType::Type(ty) => {
+                    types.type_def(ty.clone(), features, parents, offset)?;
```

Comment:
I was originally expecting to see that the `parents` list would get modified here. With component-types-in-component-types I think that recursively builds up the list of parents that can be referred back to?

Reply:
That's an excellent point and I think the spec does intend for the local type space of component types to participate in aliasing for nested component types. I'll fix.

Reply:
I've pushed up a refactoring such that component and instance type index spaces are pushed onto the list of all parent component index spaces, allowing nested component/instance types to alias their outer type.